### PR TITLE
Implement AttachmentCollectionResource

### PR DESCRIPTION
**Pull request by furinvader**
# Subset of #4768
## To-Do
- [x] Implement base version of the `AttachmentCollectionResource`
- [x] Make `attachments` property of a work package a `AttachmentCollectionResource`
- [x] Implement `opFileUpload` service
- [x] Move the functionality of the `wpAttachments` service to `AttachmentCollectionResource` 
  and `WorkPackageResource`
  - [x] Add `uploadAttachments` method to `WorkPackageResource`, that handles everything concerning the attachments after the upload
  - [x] Remove the `wpAttachments` service and its usages.
- [x] Fix file upload on work package creation
  - [x] Fix showing the uploaded file after work package creation
- [x] Fix specs
- [x] The attachments are displayed twice after uploading. This happens for newly created work packages as well as for editing existing ones. Upon reload, only one is displayed.
## Out of Scope / Dismissed
- ~~Refactor `halResourceTypes` service tests~~
- ~~Add override-child-type configuration to `halResourceTypes`~~

### CODE REVIEW

File: frontend/app/components/api/op-file-upload/op-file-upload.service.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import IQService = angular.IQService;
+import IPromise = angular.IPromise;
+
+export interface UploadFile extends File {
+  description: string;
+}
+
+export interface UploadResult {
+  uploads: IPromise<any>[];
+  upload: IPromise<any>;
```
Comment (ulferts) [inline frontend/app/components/api/op-file-upload/op-file-upload.service.ts]: I believe this should rather be called `uploaded` or `finished`

File: frontend/app/components/wp-attachments/wp-attachments.directive.ts [hunk 2]
```diff
@@ -143,24 +147,23 @@ export class WorkPackageAttachmentsController {
   }
 
   public remove(file):void {
-    if (!this.workPackage.isNew) {
-      if (file._type === 'Attachment') {
```
Comment (ulferts) [inline frontend/app/components/wp-attachments/wp-attachments.directive.ts:147]: Thanks for merging those two statements 👍

**Comment by ulferts** [comment]
File upload does not work upon work package creation. (added ToDo)

**Comment by ulferts** [comment]
There are two bugs in here one of which was already present before:
1) present before - there is no upload notification displayed when an attachment is uploaded to a newly created work package
2) new - the attachment uploaded to a newly created work package is only displayed after one reloads or presses the "Show all attributes" button.

I think at least 2 should be fixed before merging, ideally also 1 if feasible.

**Comment by ulferts** [comment]
Todo about displaying uploaded filed upon wp creation added.

**Comment by furinvader** [comment]
Could not fix 1)
We should fix this, though, after the attachment button is done.

**Comment by ulferts** [comment]
Now, the attachments are displayed twice after uploading. This happens for newly created work packages as well as for editing existing once. Upon reload, only one is displayed. 

Added ToDo.

**Comment by furinvader** [comment]
Thanks for your feedback. It's fixed now.

**Comment by ulferts** [comment]
Good to merge upon Travis' consent

**Comment by ulferts** [comment]
Thanks for the work @furinvader

**Comment by furinvader** [comment]
As for the first bug:
The notifications service automatically removes older notifications as soon as new ones appear.

I think this behaviour is faulty, because important information may be lost when multiple operations have to emit notifications.

**Pull request merged**